### PR TITLE
Documented Bulk IO issues

### LINKED ISSUE #2318: specify that access keys need to be URI encoded
I've personally hit this issue, and another user of ours has as well. Currently, when you specify an access key for a cloud provider in backup / restore, you have to URI encode special characters. Clarify this in the docs to reduce the number of access issues that users hit when running backup / restore.

cc @Amruta-Ranade

### PR DESCRIPTION
**Pull request by Amruta-Ranade**
Closes #2318 #2238 #2254 #2253 #2255 

Because the PR has multiple reviewers, I have pointed out the sections to be reviewed for each reviewer. Hope that helps!

### CODE REVIEW

File: v2.0/backup.md [hunk 1]
```diff
@@ -113,15 +113,26 @@ The URL for your backup's destination/locations must use the following format:
...
+- The backup location parameters often contain special characters that need to be URI-encoded. Use Javascript's [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) function or Go language's [url.QueryEscape](https://golang.org/pkg/net/url/#QueryEscape) function to URI-encode the parameters. Other languages provide similar functions to URI-encode special characters.
+- [<sup>1</sup>](#import-file-temp-directory-urls) Because CockroachDB is a distributed system, you cannot meaningfully store backups "locally" on nodes. The entire backup file must be stored in a single location, so attempts to store backups locally must point to an NFS drive to be useful.
+- [<sup>2</sup>](#import-file-temp-directory-urls) <span class="version-tag">New in v2.0</span> The file system backup location on a local node or NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the flag is set to `disabled`, then backups to local directories and NFS drives are disabled.
```
Comment (Amruta-Ranade) [inline v2.0/backup.md]: @dt Please review this section.

File: v2.0/backup.md [hunk 2]
```diff
@@ -113,15 +113,26 @@ The URL for your backup's destination/locations must use the following format:
...
 | HTTP | `http` | Remote host | N/A |
-| NFS | `nodelocal` | File system location | N/A |
+| NFS [<sup>1</sup>](#considerations) | `nodelocal` | File system location [<sup>2</sup>](#considerations) | N/A |
+| S3-compatible services | `s3` | Bucket name | `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `AWS_ENDPOINT` |
+
+#### Considerations
 
-{{site.data.alerts.callout_info}}Because CockroachDB is a distributed system, you cannot meaningfully store backups "locally" on nodes. The entire backup file must be stored in a single location, so attempts to store backups locally must point to an NFS drive to be useful.{{site.data.alerts.end}}
+- The backup location parameters often contain special characters that need to be URI-encoded. Use Javascript's [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) function or Go language's [url.QueryEscape](https://golang.org/pkg/net/url/#QueryEscape) function to URI-encode the parameters. Other languages provide similar functions to URI-encode special characters.
```
Comment (Amruta-Ranade) [inline v2.0/backup.md]: @benesch Please review this section

File: v2.0/import.md [hunk 3]
```diff
@@ -172,14 +172,17 @@ URLs for the file you want to import and your temp directory must use the follow
 | Google Cloud [<sup>1</sup>](#notes) | `gs` | Bucket name | None |
 | HTTP [<sup>2</sup>](#notes) | `http` | Remote host | N/A |
 | NFS/Local [<sup>3</sup>](#notes) | `nodelocal` | File system location | N/A |
+| S3-compatible services | `s3` | Bucket name | `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `AWS_ENDPOINT` |
```
Comment (Amruta-Ranade) [inline v2.0/import.md]: @mjibson Please review this row.

File: v2.0/restore.md [hunk 4]
```diff
@@ -109,8 +109,13 @@ The URL for your backup's locations must use the following format:
 | Google Cloud Storage | `gs` | Bucket name | None––currently only supports instance auth, but we can build non-instance auth at a customer's request |
 | HTTP | `http` | Remote host | N/A |
 | NFS | `nodelocal` | File system location | N/A |
+| S3-compatible services | `s3` | Bucket name | `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `AWS_ENDPOINT` |
```
Comment (Amruta-Ranade) [inline v2.0/restore.md]: @mjibson Please review this row.

File: v2.0/restore.md [hunk 5]
```diff
@@ -109,8 +109,13 @@ The URL for your backup's locations must use the following format:
...
 
-{{site.data.alerts.callout_info}}Backups stored on NFSes work only if each node in the cluster accesses the drive in the same way.{{site.data.alerts.end}}
+#### Considerations
+
+- The backup location parameters often contain special characters that need to be URI-encoded. Use Javascript's [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) function or Go language's [url.QueryEscape](https://golang.org/pkg/net/url/#QueryEscape) function to URI-encode the parameters. Other languages provide similar functions to URI-encode special characters.
```
Comment (Amruta-Ranade) [inline v2.0/restore.md]: @benesch Please review this section

File: v2.0/restore.md [hunk 6]
```diff
@@ -109,8 +109,13 @@ The URL for your backup's locations must use the following format:
...
+- Backups stored on NFSes work only if each node in the cluster accesses the drive in the same way.
```
Comment (Amruta-Ranade) [inline v2.0/restore.md]: @dt Please review this section

File: v2.0/start-a-node.md [hunk 7]
```diff
@@ -42,6 +42,7 @@ Flag | Description
...
 `--certs-dir` | The path to the [certificate directory](create-security-certificates.html). The directory must contain valid certificates if running in secure mode.<br><br>**Default:** `${HOME}/.cockroach-certs/`
+`--external-io-dir` | <span class="version-tag">New in v2.0</span> The path of the external IO directory with which the local file access paths are prefixed while performing backup and restore operations using local node directories or NFS drives. If set to `disabled`, backups and restores using local node directories and NFS drives are disabled.<br><br>**Default:** `extern` directory on the first storage device specified by the `--store` flag.
```
Comment (Amruta-Ranade) [inline v2.0/start-a-node.md]: @dt Please review this section

File: v2.0/backup.md [hunk 8]
```diff
@@ -113,15 +113,26 @@ The URL for your backup's destination/locations must use the following format:
...
+
 | Backup Location | scheme | host | parameters |
 |-----------------|--------|------|------------|
 | Amazon S3 | `s3` | Bucket name | `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` |
 | Azure | `azure` | Container name | `AZURE_ACCOUNT_KEY`, `AZURE_ACCOUNT_NAME` |
 | Google Cloud Storage | `gs` | Bucket name | None––currently only supports instance auth, but we can build non-instance auth at a customer's request |
 | HTTP | `http` | Remote host | N/A |
-| NFS | `nodelocal` | File system location | N/A |
+| NFS [<sup>1</sup>](#considerations) | `nodelocal` | File system location [<sup>2</sup>](#considerations) | N/A |
+| S3-compatible services | `s3` | Bucket name | `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `AWS_ENDPOINT` |
```
Comment (Amruta-Ranade) [inline v2.0/backup.md]: @mjibson Please review this row.

File: v2.0/import.md [hunk 9]
```diff
@@ -172,14 +172,17 @@ URLs for the file you want to import and your temp directory must use the follow
...
 
 #### Notes
 
 [<sup>1</sup>](#import-file-temp-directory-urls) Only supports instance auth.
 
 [<sup>2</sup>](#import-file-temp-directory-urls) You can easily create your own HTTP server with [Caddy or nginx](create-a-file-server.html).
 
-[<sup>3</sup>](#import-file-temp-directory-urls) If using NFS for your temp directory, each node in the cluster must have access to the NFS using the same URL.
+[<sup>3</sup>](#import-file-temp-directory-urls) If using NFS for your temp directory, each node in the cluster must have access to the NFS using the same URL. The file system backup location on the NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the flag is set to `disabled`, then imports from local directories and NFS drives are disabled.
```
Comment (Amruta-Ranade) [inline v2.0/import.md]: @dt Please review this section

File: v2.0/import.md [hunk 10]
```diff
@@ -172,14 +172,17 @@ URLs for the file you want to import and your temp directory must use the follow
...
 
 [<sup>2</sup>](#import-file-temp-directory-urls) You can easily create your own HTTP server with [Caddy or nginx](create-a-file-server.html).
 
-[<sup>3</sup>](#import-file-temp-directory-urls) If using NFS for your temp directory, each node in the cluster must have access to the NFS using the same URL.
+[<sup>3</sup>](#import-file-temp-directory-urls) If using NFS for your temp directory, each node in the cluster must have access to the NFS using the same URL. The file system backup location on the NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the flag is set to `disabled`, then imports from local directories and NFS drives are disabled. 
+
+The location parameters often contain special characters that need to be URI-encoded. Use Javascript's [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) function or Go language's [url.QueryEscape](https://golang.org/pkg/net/url/#QueryEscape) function to URI-encode the parameters. Other languages provide similar functions to URI-encode special characters.
```
Comment (Amruta-Ranade) [inline v2.0/import.md]: @benesch Please review this section

**Comment by benesch** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2340#-:-L2WEdW8wHMR4Hh9XhJc:bnfp4nl)*

File: v2.0/restore.md (hunk 6 above)
Comment (dt) [reply v2.0/restore.md]: The `--external-io-dir` mitigates this limitation to some extent -- previously, since the path was relative to `/`, every node had to mount the NFS share in the same place, like `/mnt/nfs/backups` or whatever. Now that the backup path is relative to the node's configured external-io-dir, nodes can have the NFS mount at whatever location they want as long as they all point to the same share using their `external-io-dir` flag.. e.g. one node might be started with `--external-io-dir=/mnt/nfs/backups` while another might use `--external-io-dir=/Volumes/backups` or whatever.

The important thing is that all nodes use `--external-io-dir` to point to the same share.

File: v2.0/backup.md (hunk 1 above)
Comment (dt) [reply v2.0/backup.md]: This is correct, though we could include in the last one `(which defaults a subdirectory of the first configured store)` or similar if we wanted to.

File: v2.0/start-a-node.md (hunk 7 above)
Comment (dt) [reply v2.0/start-a-node.md]: I think we just call them stores? though we generally only advice configuring one per physical storage device. To be a bit pedantic, since we default to `cockroach-data`, it isn't always explicitly specified by `--store`.

I might go for something like:
>`extern` subdirectory of the first configured store (see the `--store` flag).

File: v2.0/import.md (hunk 9 above)
Comment (dt) [reply v2.0/import.md]: First sentence of this is no longer quite true (due to the second).
Now it is important only that each node in your cluster configure its `--external-io-dir` flag to point to its local mount point of the that NFS share (or, as an alternative, symlink from the default `extern` subdirectory to the NFS-backed directory).

File: v2.0/backup.md (hunk 1 above)
Comment (dt) [reply v2.0/backup.md]: We could also mention that, if you do not want to restart your nodes to change this flag, you can create symlinks to the locations you want to use from within the extern dir to read/write to paths outside the extern dir

Comment (Amruta-Ranade) [reply v2.0/backup.md]: @dt Does this sound right: "The file system backup location on a local node or NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the `--external-io-dir` flag is set to `disabled`, then backups to local directories and NFS drives are disabled. By default, the flag is set to the `extern` subdirectory of the first configured [`store`](#store). To avoid restarting your nodes to set the `--external-io-dir` flag to the locations you want to use, create symlinks to the desired locations from within the `extern` directory."

File: v2.0/restore.md (hunk 6 above)
Comment (Amruta-Ranade) [reply v2.0/restore.md]: @dt So I can delete the second bullet (Backups stored on NFSes...), right?

**Comment by dt** [comment]
Review status: 1 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[v2.0/backup.md, line 135 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2340#-L2RWu0J4iki_CTyXMDc-r1-135:-L2zcl5LOo2yOfKhRgo2:b-i7544q) ([raw file](https://github.com/cockroachdb/docs/blob/c305a6b4ddec749e364b266e72862bbac4c74bfe/v2.0/backup.md#L135)):*
<details><summary><i>Previously, Amruta-Ranade (Amruta Ranade) wrote…</i></summary><blockquote>

@dt Does this sound right: "The file system backup location on a local node or NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the `--external-io-dir` flag is set to `disabled`, then backups to local directories and NFS drives are disabled. By default, the flag is set to the `extern` subdirectory of the first configured [`store`](#store). To avoid restarting your nodes to set the `--external-io-dir` flag to the locations you want to use, create symlinks to the desired locations from within the `extern` directory."
</blockquote></details>

:thumbsup: that last sentence could be prefixed with `Note:` or something or have a `you can` or some indication that it is an optional suggestion, not a strong recommendation.

---

*[v2.0/restore.md, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2340#-L2RWu0J4iki_CTyXMDe-r1-117:-L2zd7g_GBXNWX9ekPoH:bb7sa0c) ([raw file](https://github.com/cockroachdb/docs/blob/c305a6b4ddec749e364b266e72862bbac4c74bfe/v2.0/restore.md#L117)):*
<details><summary><i>Previously, Amruta-Ranade (Amruta Ranade) wrote…</i></summary><blockquote>

@dt So I can delete the second bullet (Backups stored on NFSes...), right?
</blockquote></details>

yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2340)*

**Comment by dt** [comment]
:lgtm:

---

Review status: 1 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2340#-:-L2zdCh6PB-1I4x6uwvb:bnfp4nl)*

File: v2.0/backup.md [hunk 11]
```diff
@@ -113,15 +113,26 @@ The URL for your backup's destination/locations must use the following format:
...
+- [<sup>1</sup>](#import-file-temp-directory-urls) Because CockroachDB is a distributed system, you cannot meaningfully store backups "locally" on nodes. The entire backup file must be stored in a single location, so attempts to store backups locally must point to an NFS drive to be useful.
+- [<sup>2</sup>](#import-file-temp-directory-urls) <span class="version-tag">New in v2.0</span> The file system backup location on a local node or NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the `--external-io-dir` flag is set to `disabled`, then backups to local directories and NFS drives are disabled. By default, the flag is set to the `extern` subdirectory of the first configured [`store`](#store). To avoid restarting your nodes to set the `--external-io-dir` flag to the locations you want to use, create symlinks to the desired locations from within the `extern` directory.
```
Comment (jseldess) [inline v2.0/backup.md]: - The `<span class="version-tag">New in v2.0</span>` doesn't behave well will bulleted lists. Renders weird:

    screen shot 2018-01-19 at 10 11 27 am

    So let's just remove it here. It's called out as new in the `cockroach start` docs, which is sufficient. 

- `(#store)` should be `(start-a-node.html#store)`.

- In the last sentence, "To avoid restarting your nodes to set the `--external-io-dir` flag to the locations you want to use` is confusing me. Can you rephrase?

File: v2.0/import.md [hunk 12]
```diff
@@ -172,14 +172,17 @@ URLs for the file you want to import and your temp directory must use the follow
 | Google Cloud [<sup>1</sup>](#notes) | `gs` | Bucket name | None |
 | HTTP [<sup>2</sup>](#notes) | `http` | Remote host | N/A |
 | NFS/Local [<sup>3</sup>](#notes) | `nodelocal` | File system location | N/A |
+| S3-compatible services | `s3` | Bucket name | `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, `AWS_ENDPOINT` |
 
 #### Notes
```
Comment (jseldess) [inline v2.0/import.md]: Since this is called `Considerations` in `backup.md`, perhaps use that heading here as well? Also use bullets like in `backup.md`?

File: v2.0/restore.md (hunk 6 above)
Comment (jseldess) [reply v2.0/restore.md]: @Amruta-Ranade, @dt, has this conversation been resolved?

File: v2.0/restore.md [hunk 13]
```diff
@@ -109,8 +109,13 @@ The URL for your backup's locations must use the following format:
...
+- <span class="version-tag">New in v2.0</span> The file system backup location on a local node or NFS drive is relative to the path specified by the `--external-io-dir` flag set while [starting the node](start-a-node.html). If the flag is set to `disabled`, then backups and restores from local directories and NFS drives are disabled.
```
Comment (jseldess) [inline v2.0/restore.md]: Same feedback as for `backup.md`: Let's remove `<span class="version-tag">New in v2.0</span>` here.

File: v2.0/start-a-node.md [hunk 14]
```diff
@@ -42,6 +42,7 @@ Flag | Description
...
 `--certs-dir` | The path to the [certificate directory](create-security-certificates.html). The directory must contain valid certificates if running in secure mode.<br><br>**Default:** `${HOME}/.cockroach-certs/`
+`--external-io-dir` | <span class="version-tag">New in v2.0</span> The path of the external IO directory with which the local file access paths are prefixed while performing backup and restore operations using local node directories or NFS drives. If set to `disabled`, backups and restores using local node directories and NFS drives are disabled.<br><br>**Default:** `extern` subdirectory of the first configured `store` (see [`store`](#store) below).<br><br>To avoid restarting your nodes to set the `--external-io-dir` flag to the locations you want to use, create symlinks to the desired locations from within the `extern` directory.
```
Comment (jseldess) [inline v2.0/start-a-node.md]: - Add colon to "new" callout, i.e., `<span class="version-tag">New in v2.0:</span>`.
- Let's remove "see `store` below" and just make the previous instance of `store` a link, i.e.,
    > the first configured [`store`](#store).
- As I mentioned in my review of `backup.md`, I don't understand what this is saying:  
    > To avoid restarting your nodes to set the `--external-io-dir` flag to the locations you want to use, create symlinks to the desired locations from within the `extern` directory. 

    Can you rephrase, or explain it to me in other terms so I can figure out what I'm not getting?

**Comment by jseldess** [comment]
LGTM

**Comment by jseldess** [comment]
TC is failing on a link to www.bottlecaps.de/rr/ui, which is an attribution at the bottom of our full sql grammar page. Merging in the expectation that the site comes back eventually.
